airflow/migrations/versions/0041_3_0_0_add_uuid_primary_key_to_task_instance_.py

```diff
@@ -0,0 +1,273 @@
+    SET time_hex = LPAD(HEX(unix_time_ms), 12, '0');
+    SET rand_hex = CONCAT(
+        LPAD(HEX(FLOOR(RAND() * POW(2,32))), 8, '0'),
+        LPAD(HEX(FLOOR(RAND() * POW(2,32))), 8, '0')
+    );
+    SET rand_hex = CONCAT(SUBSTRING(rand_hex, 1, 4), '7', SUBSTRING(rand_hex, 6));
+    SET rand_hex = CONCAT(SUBSTRING(rand_hex, 1, 12), '8', SUBSTRING(rand_hex, 14));
+
+    SET uuid = LOWER(CONCAT(
+        SUBSTRING(time_hex, 1, 8), '-',
+        SUBSTRING(time_hex, 9, 4), '-',
+        SUBSTRING(rand_hex, 1, 4), '-',
+        SUBSTRING(rand_hex, 5, 4), '-',
+        SUBSTRING(rand_hex, 9)
+    ));
+
+    RETURN uuid;
+END;
+"""
+
+mysql_uuid7_fn_drop = """
+DROP FUNCTION IF EXISTS uuid_generate_v7;
+"""
+
+ti_table = "task_instance"
+
+# Foreign key columns from task_instance
+ti_fk_cols = ["dag_id", "task_id", "run_id", "map_index"]
+
+# Foreign key constraints from other tables to task_instance
+ti_fk_constrains = [
+    {"table": "rendered_task_instance_fields", "fk": "rtif_ti_fkey"},
+    {"table": "task_fail", "fk": "task_fail_ti_fkey"},
+    {"table": "task_instance_history", "fk": "task_instance_history_ti_fkey"},
+    {"table": "task_instance_note", "fk": "task_instance_note_ti_fkey"},
+    {"table": "task_map", "fk": "task_map_task_instance_fkey"},
+    {"table": "task_reschedule", "fk": "task_reschedule_ti_fkey"},
+    {"table": "xcom", "fk": "xcom_task_instance_fkey"},
+]
+
+
+def _get_type_id_column(dialect_name: str) -> sa.types.TypeEngine:
+    # For PostgreSQL, use the UUID type directly as it is more efficient
+    if dialect_name == "postgresql":
+        return postgresql.UUID(as_uuid=False)
+    # For other databases, use String(36) to match UUID format
+    else:
+        return sa.String(36)
+
+
+def upgrade():
+    """Add UUID primary key to task instance table."""
+    conn = op.get_bind()
+    dialect_name = conn.dialect.name
+
+    op.add_column("task_instance", sa.Column("id", _get_type_id_column(dialect_name), nullable=True))
+
+    if dialect_name == "postgresql":
+        op.execute(pg_uuid7_fn)
+
+        # TODO: Add batching to handle updates in smaller chunks for large tables to avoid locking
+        # Migrate existing rows with UUID v7 using a timestamp-based generation
+        op.execute(
+            "UPDATE task_instance SET id = uuid_generate_v7(coalesce(queued_dttm, start_date, clock_timestamp()))"
+        )
+
+        op.execute(pg_uuid7_fn_drop)
+
+        # Drop existing primary key constraint to task_instance table
+        op.execute("ALTER TABLE IF EXISTS task_instance DROP CONSTRAINT task_instance_pkey CASCADE")
+
+    elif dialect_name == "mysql":
+        op.execute(mysql_uuid7_fn)
+
+        # Migrate existing rows with UUID v7
+        op.execute("""
+            UPDATE task_instance
+            SET id = uuid_generate_v7(coalesce(queued_dttm, start_date, NOW(3)))
+            WHERE id IS NULL
+        """)
+
+        op.execute(mysql_uuid7_fn_drop)
```

Comment:
So these functions are used exclusively in the migrations? Might be worth adding a comment to make that clear.

Reply:
Good point. Done in https://github.com/apache/airflow/pull/43243/commits/0f5dc0a0e496a38d53003ee9dc2941af0d8c9b74